File: apps/web/src/models/community.ts
/* ─────────────────────────────────────────────────────
   Community & Linked-member data models

   NOTE: These are the local/UI models used by components.
   The API layer types (AuthorOut, PostListOut, etc.) live
   in src/lib/api/community.ts. These models add compat
   helpers and local-only fields.
───────────────────────────────────────────────────── */

import type { AuthorOut, Attachment } from "@/lib/api/community";

export type PostReply = {
  id?: number;
  /** Author display name (string for direct JSX rendering) */
  author?: string;
  /** Structured author object from API — use for avatar, role, etc. */
  authorData?: AuthorOut;
  initials: string;
  authorName?: string;
  time: string;
  text: string;
  attachments?: Attachment[];
  likes?: number;
  created_at?: string;
};

export type CommunityPost = {
  id: number;
  group_id?: string;
  /** New: structured author from API */
  authorObj?: AuthorOut;
  /** Legacy flat fields — kept for backward compat */
  author: string;
  initials: string;
  location: string;
  time: string;
  text: string;
  likes: number;
  replyCount: number;
  tag: string;
  attachments?: Attachment[];
  is_pinned?: boolean;
  created_at?: string;
  /** Replies are NOT included in list responses — fetched on-demand per post */
  replies?: PostReply[];
};

export type LinkedPost = {
  id: number;
  /** New: structured author from API */
  authorObj?: AuthorOut;
  /** Legacy flat fields */
  initials: string;
  author: string;
  time: string;
  text: string;
  likes: number;
  replyCount: number;
  tag: string;
  attachments?: Attachment[];
  /** Replies are NOT included in list responses — fetched on-demand per post */
  replies?: PostReply[];
};

export type LinkedMember = {
  name: string;
  relation: string;
  direction: string;
  scope: string;
  badgeLabel: string;
  initials: string;
  posts: LinkedPost[];
  sharedFiles?: { name: string; size: string; date: string }[];
  memberCount?: number | string;
  members?: { name: string; role: string; initials: string; status: string }[];
};

/* ── Community File types ──────────────────────────────────────── */

export type FileQA = {
  id: number;
  question: string;
  askedBy: string;
  askedByInitials: string;
  askedAt: string;
  answer: string;
};

export type CommunityFile = {
  id: number;
  name: string;
  type: "pdf" | "xlsx" | "docx" | "jpg" | "png";
  size: string;
  uploadedBy: string;
  uploadedByInitials: string;
  uploadedAt: string;
  category: string;
  aiSummary: string;
  qaCount: number;
  questions: FileQA[];
};

/* ── Member Activity types ─────────────────────────────────────── */

export type MemberActivityType = "post" | "reply" | "upload" | "question" | "like";

export type MemberActivity = {
  id: number;
  type: MemberActivityType;
  time: string;
  text: string;
  context?: string;
  tag?: string;
};

export type CommunityMember = {
  id: number;
  name: string;
  initials: string;
  role: string;
  status: "online" | "recently" | "offline";
  statusLabel: string;
  joinedAt: string;
  location?: string;
  stats: {
    posts: number;
    replies: number;
    uploads: number;
    questions: number;
    likes: number;
  };
  activities: MemberActivity[];
};

/* ── Helpers: map API response → local model ───────────────────── */

/**
 * Convert API PostListOut (with AuthorOut) → flat CommunityPost
 * for backward compat with components that read post.author, post.initials.
 */
export function apiPostToLocal(
  apiPost: {
    id: number;
    group_id: string;
    author: AuthorOut;
    location: string;
    time: string;
    text: string;
    likes: number;
    replyCount: number;
    tag: string;
    attachments?: Attachment[];
    is_pinned?: boolean;
    created_at?: string;
  },
): CommunityPost {
  return {
    id: apiPost.id,
    group_id: apiPost.group_id,
    authorObj: apiPost.author,
    author: apiPost.author.name,
    initials: apiPost.author.initials,
    location: apiPost.location,
    time: apiPost.time,
    text: apiPost.text,
    likes: apiPost.likes,
    replyCount: apiPost.replyCount,
    tag: apiPost.tag,
    attachments: apiPost.attachments ?? [],
    is_pinned: apiPost.is_pinned ?? false,
    created_at: apiPost.created_at ?? undefined,
  };
}

/**
 * Convert API PostReplyOut (with AuthorOut) → flat PostReply
 */
export function apiReplyToLocal(
  apiReply: {
    id: number;
    author: AuthorOut;
    time: string;
    text: string;
    attachments?: Attachment[];
    likes?: number;
    created_at?: string;
  },
): PostReply {
  return {
    id: apiReply.id,
    author: apiReply.author.name,
    authorData: apiReply.author,
    initials: apiReply.author.initials,
    authorName: apiReply.author.name,
    time: apiReply.time,
    text: apiReply.text,
    attachments: apiReply.attachments ?? [],
    likes: apiReply.likes ?? 0,
    created_at: apiReply.created_at ?? undefined,
  };
}

File: apps/backend/app/api/schemas/community.py
"""
Pydantic schemas for community endpoints.
Mirrors shared TS types from apps/web/src/models/community.ts.

Industry-standard patterns:
  - Cursor-based pagination for feeds (scalable, no offset drift)
  - Offset pagination for bounded lists (files, members)
  - Rich media attachments for posts & replies (images, voice, documents)
  - Consistent response envelopes
"""

from __future__ import annotations

from pydantic import BaseModel, Field
from enum import Enum


# ══════════════════════════════════════════════════════════════════════════════
#  ENUMS
# ══════════════════════════════════════════════════════════════════════════════


class AttachmentType(str, Enum):
    IMAGE = "image"
    VOICE = "voice"
    DOCUMENT = "document"


class PostSortField(str, Enum):
    CREATED = "created"
    LIKES = "likes"
    REPLIES = "replies"


class SortOrder(str, Enum):
    ASC = "asc"
    DESC = "desc"


# ══════════════════════════════════════════════════════════════════════════════
#  PAGINATION
# ══════════════════════════════════════════════════════════════════════════════


class CursorPageMeta(BaseModel):
    """Cursor-based pagination metadata — ideal for feeds."""
    has_next: bool = False
    next_cursor: str | None = None
    has_prev: bool = False
    prev_cursor: str | None = None
    total: int = 0


class OffsetPageMeta(BaseModel):
    """Offset-based pagination metadata — ideal for bounded lists."""
    page: int = 1
    page_size: int = 20
    total: int = 0
    total_pages: int = 0


# ══════════════════════════════════════════════════════════════════════════════
#  ATTACHMENTS — rich media for posts & replies
# ══════════════════════════════════════════════════════════════════════════════


class Attachment(BaseModel):
    """A media attachment on a post or reply."""
    id: str
    type: AttachmentType
    url: str  # signed URL or relative path
    filename: str | None = None
    mime_type: str | None = None
    size_bytes: int | None = None
    duration_seconds: float | None = None  # voice recordings
    thumbnail_url: str | None = None  # image/document preview
    transcription: str | None = None  # voice → text
    metadata: dict | None = None


class AttachmentCreate(BaseModel):
    """Client sends this when creating a post/reply with attachments."""
    type: AttachmentType
    url: str
    filename: str | None = None
    mime_type: str | None = None
    size_bytes: int | None = None
    duration_seconds: float | None = None
    thumbnail_url: str | None = None
    transcription: str | None = None


# ══════════════════════════════════════════════════════════════════════════════
#  POSTS
# ══════════════════════════════════════════════════════════════════════════════


class AuthorOut(BaseModel):
    """Author info embedded in posts/replies."""
    id: str
    name: str
    initials: str
    role: str | None = None
    avatar_url: str | None = None


class PostReplyOut(BaseModel):
    id: int
    author: AuthorOut
    time: str
    text: str
    attachments: list[Attachment] = []
    likes: int = 0
    created_at: str | None = None

    # Keep flat initials/author for backward compat
    @property
    def initials(self) -> str:
        return self.author.initials

    @property
    def author_name(self) -> str:
        return self.author.name


class PostListOut(BaseModel):
    """Post in list view — no replies, lightweight."""
    id: int
    group_id: str
    author: AuthorOut
    location: str
    time: str
    text: str
    likes: int = 0
    replyCount: int = 0
    tag: str
    attachments: list[Attachment] = []
    is_pinned: bool = False
    created_at: str | None = None


class PostOut(BaseModel):
    """Full post detail with embedded replies."""
    id: int
    group_id: str
    author: AuthorOut
    location: str
    time: str
    text: str
    likes: int = 0
    replyCount: int = 0
    tag: str
    attachments: list[Attachment] = []
    replies: list[PostReplyOut] = []
    is_pinned: bool = False
    created_at: str | None = None


class PostCreate(BaseModel):
    text: str = Field(..., min_length=1, max_length=5000)
    tag: str = "Discussion"
    attachments: list[AttachmentCreate] = []


class ReplyCreate(BaseModel):
    text: str = Field(..., min_length=1, max_length=2000)
    attachments: list[AttachmentCreate] = []


class LikeToggleResponse(BaseModel):
    post_id: int
    liked: bool
    total_likes: int


# ── Paginated responses ──────────────────────────────────────────────────────

class PostListResponse(BaseModel):
    """Cursor-paginated post feed."""
    items: list[PostListOut]
    meta: CursorPageMeta


class ReplyListResponse(BaseModel):
    """Offset-paginated replies for a post."""
    items: list[PostReplyOut]
    meta: OffsetPageMeta


# ══════════════════════════════════════════════════════════════════════════════
#  AI — summaries, rephrasings
# ══════════════════════════════════════════════════════════════════════════════


class PostSummaryOut(BaseModel):
    post_id: int
    summary: str
    ai_response: str


class RephraseRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=2000)


class RephraseResponse(BaseModel):
    original: str
    rephrasings: list[str]


# ══════════════════════════════════════════════════════════════════════════════
#  FILES
# ══════════════════════════════════════════════════════════════════════════════


class FileQAOut(BaseModel):
    id: int
    question: str
    askedBy: str
    askedByInitials: str
    askedAt: str
    answer: str


class FileOut(BaseModel):
    id: int
    group_id: str
    name: str
    type: str
    size: str
    uploadedBy: str
    uploadedByInitials: str
    uploadedAt: str
    category: str
    aiSummary: str
    qaCount: int = 0
    questions: list[FileQAOut] = []


class FileUploadOut(BaseModel):
    """Response after uploading a file."""
    id: int
    group_id: str
    name: str
    type: str
    size: str
    uploadedBy: str
    uploadedByInitials: str
    uploadedAt: str
    category: str
    aiSummary: str


class FileQACreate(BaseModel):
    question: str = Field(..., min_length=1, max_length=1000)


class RecentFileQAOut(BaseModel):
    fileId: int
    fileName: str
    fileCategory: str
    question: str
    askedBy: str
    askedByInitials: str
    askedAt: str
    answer: str


class FileListResponse(BaseModel):
    """Offset-paginated file list."""
    items: list[FileOut]
    meta: OffsetPageMeta


# ══════════════════════════════════════════════════════════════════════════════
#  MEMBERS
# ══════════════════════════════════════════════════════════════════════════════


class MemberActivityOut(BaseModel):
    id: int
    type: str  # "post" | "reply" | "upload" | "question" | "like"
    time: str
    text: str
    context: str | None = None
    tag: str | None = None


class MemberStatsOut(BaseModel):
    posts: int = 0
    replies: int = 0
    uploads: int = 0
    questions: int = 0
    likes: int = 0


class MemberOut(BaseModel):
    id: int
    name: str
    initials: str
    role: str
    status: str  # "online" | "recently" | "offline"
    statusLabel: str
    joinedAt: str
    location: str | None = None
    stats: MemberStatsOut
    activities: list[MemberActivityOut] = []


class MemberListResponse(BaseModel):
    """Offset-paginated member list."""
    items: list[MemberOut]
    meta: OffsetPageMeta
